File: beams/beams/custom_scripts/budget/budget.js
frappe.ui.form.on('Budget', {
    refresh: function (frm) {
        set_filters(frm);
        if (!frm.is_new()) {
            frm.add_custom_button('Open Budget Tool', () => {
                frappe.set_route('Form', 'Budget Tool', 'Budget Tool');
            });
        }
    },
    department: function (frm) {
        set_filters(frm);
        if (!frm.doc.department) {
            frm.set_value('division', null);
        }
    },
    company: function (frm) {
        frm.set_value('department', null);
    },
    budget_template: function (frm) {
    if (!frm.doc.budget_template) {
        frm.set_value('cost_center', null);
        frm.set_value('region', null);
        frm.clear_table('accounts');
        frm.refresh_field('accounts');
        return;
    }

    if (frm.doc.budget_template === frm._previous_budget_template) {
        return;
    }

    let previous_template = frm.doc.__last_value || frm._previous_budget_template;

    frappe.confirm(
        __('Are you sure you want to change the Budget Template? This will reset existing budget data.'),
        function () {
            frm.clear_table('accounts');

            frappe.call({
                method: 'frappe.client.get',
                args: {
                    doctype: 'Budget Template',
                    name: frm.doc.budget_template
                },
                callback: function (response) {
                    if (response.message) {
                        let budget_template = response.message;
                        frm.set_value('cost_center', budget_template.cost_center);
                        frm.set_value('region', budget_template.region);

                        let budget_template_items = budget_template.budget_template_item || [];
                        budget_template_items.forEach(function (item) {
                            let row = frm.add_child('accounts');
                            row.cost_head = item.cost_head;
                            row.cost_subhead = item.cost_sub_head;
                            row.account = item.account;
                            row.cost_category = item.cost_category;
                        });
                        frm.refresh_field('accounts');
                    }
                }
            });

            frm._previous_budget_template = frm.doc.budget_template;
        },
        function () {
            frm.set_value('budget_template', previous_template);
        }
    );
}
});
// Function to apply filters in the cost subhead field in Budget Account
function set_filters(frm) {
    frm.set_query('division', function () {
        return {
            filters: {
                department: frm.doc.department,
                company: frm.doc.company
            }
        };
    });
    frm.set_query('budget_template', function () {
        return {
            filters: {
                division: frm.doc.division,
                company: frm.doc.company
            }
        };
    });
    frm.set_query('department', function () {
        return {
            filters: {
                company: frm.doc.company
            }
        };
    });
    frm.set_query('region', function () {
        return {
            filters: {
                company: frm.doc.company
            }
        };
    });
}

frappe.ui.form.on('Budget Account', {
  cost_subhead: function (frm, cdt, cdn) {
      var row = locals[cdt][cdn];

      if (row.cost_subhead && frm.doc.company) {
          frappe.db.get_doc('Cost Subhead', row.cost_subhead).then(doc => {
              if (doc.accounts && doc.accounts.length > 0) {
                  let account_found = doc.accounts.find(acc => acc.company === frm.doc.company);
                  if (account_found) {
                      frappe.model.set_value(cdt, cdn, 'account', account_found.default_account);
                  } else {
                      frappe.model.set_value(cdt, cdn, 'account', '');
                      frappe.msgprint(__('No default account found for the selected Cost Subhead and Company.'));
                  }
              } else {
                  frappe.model.set_value(cdt, cdn, 'account', '');
              }
          });
      }
  },
    budget_amount: function (frm, cdt, cdn) {
        let row = locals[cdt][cdn];
        if (row.equal_monthly_distribution && row.budget_amount) {
            distribute_budget_equally(frm, cdt, cdn, row.budget_amount);
        }
    },
    equal_monthly_distribution: function (frm, cdt, cdn) {
        let row = locals[cdt][cdn];

        if (!row.equal_monthly_distribution) {
            frappe.confirm(
                "Are you sure you want to uncheck Equal Monthly Distribution?",
                function() {
                    clear_monthly_values(frm, cdt, cdn);
                },
                function() {
                    frappe.model.set_value(cdt, cdn, "equal_monthly_distribution", 1);
                }
            );
        } else if (row.budget_amount) {
            distribute_budget_equally(frm, cdt, cdn, row.budget_amount);
        }
    },
    january: function (frm, cdt, cdn) {
        calculate_budget_amount(frm, cdt, cdn);
    },
    february: function (frm, cdt, cdn) {
        calculate_budget_amount(frm, cdt, cdn);
    },
    march: function (frm, cdt, cdn) {
        calculate_budget_amount(frm, cdt, cdn);
    },
    april: function (frm, cdt, cdn) {
        calculate_budget_amount(frm, cdt, cdn);
    },
    may: function (frm, cdt, cdn) {
        calculate_budget_amount(frm, cdt, cdn);
    },
    june: function (frm, cdt, cdn) {
        calculate_budget_amount(frm, cdt, cdn);
    },
    july: function (frm, cdt, cdn) {
        calculate_budget_amount(frm, cdt, cdn);
    },
    august: function (frm, cdt, cdn) {
        calculate_budget_amount(frm, cdt, cdn);
    },
    september: function (frm, cdt, cdn) {
        calculate_budget_amount(frm, cdt, cdn);
    },
    october: function (frm, cdt, cdn) {
        calculate_budget_amount(frm, cdt, cdn);
    },
    november: function (frm, cdt, cdn) {
        calculate_budget_amount(frm, cdt, cdn);
    },
    december: function (frm, cdt, cdn) {
        calculate_budget_amount(frm, cdt, cdn);
    }
});

function calculate_budget_amount(frm, cdt, cdn) {
    let row = locals[cdt][cdn];
    // Calculate the total of all monthly amounts
    let total =
        (row.january || 0) +
        (row.february || 0) +
        (row.march || 0) +
        (row.april || 0) +
        (row.may || 0) +
        (row.june || 0) +
        (row.july || 0) +
        (row.august || 0) +
        (row.september || 0) +
        (row.october || 0) +
        (row.november || 0) +
        (row.december || 0);

    frappe.model.set_value(cdt, cdn, 'budget_amount', total);
    frm.refresh_field('budget_account');
}

function distribute_budget_equally(frm, cdt, cdn, budget_amount) {
    let row = locals[cdt][cdn];

    // Calculate equal amount for each month and rounding adjustment
    let equal_amount = Math.floor((budget_amount / 12) * 100) / 100;
    let total = equal_amount * 12;
    let difference = Math.round((budget_amount - total) * 100) / 100;

    // Distribute the amounts
    frappe.model.set_value(cdt, cdn, 'january', equal_amount);
    frappe.model.set_value(cdt, cdn, 'february', equal_amount);
    frappe.model.set_value(cdt, cdn, 'march', equal_amount);
    frappe.model.set_value(cdt, cdn, 'april', equal_amount);
    frappe.model.set_value(cdt, cdn, 'may', equal_amount);
    frappe.model.set_value(cdt, cdn, 'june', equal_amount);
    frappe.model.set_value(cdt, cdn, 'july', equal_amount);
    frappe.model.set_value(cdt, cdn, 'august', equal_amount);
    frappe.model.set_value(cdt, cdn, 'september', equal_amount);
    frappe.model.set_value(cdt, cdn, 'october', equal_amount);
    frappe.model.set_value(cdt, cdn, 'november', equal_amount);
    frappe.model.set_value(cdt, cdn, 'december', equal_amount + difference);

    frm.refresh_field('budget_account');
}

function clear_monthly_values(frm, cdt, cdn) {
    let fields = [
        'january', 'february', 'march', 'april', 'may', 'june',
        'july', 'august', 'september', 'october', 'november', 'december'
    ];

    fields.forEach(field => frappe.model.set_value(cdt, cdn, field, 0));

    frm.refresh_field('budget_account');
}

frappe.ui.form.on("Rejection Feedback", {
    rejection_feedback_add: function(frm, cdt, cdn) {
        let row = frappe.get_doc(cdt, cdn);
        row.user = frappe.session.user_fullname;
        frm.refresh_field("rejection_feedback");
    }
});

File: beams/beams/custom_scripts/budget/budget.py
import frappe

month_fields = ['january', 'february', 'march', 'april', 'may', 'june', 'july', 'august', 'september', 'october', 'november', 'december']

def beams_budget_validate(doc, method=None):
	'''
		Method trigger on `validate` event of Budget
	'''
	update_budget_against(doc, method)
	update_total_amount(doc, method)
	convert_currency(doc, method)

def update_total_amount(doc, method):
	total = sum([row.budget_amount for row in doc.get('budget_accounts') if row.budget_amount])
	doc.total_amount = total

def populate_og_accounts(doc, method=None):
	'''
		Method trigger on `before_validate` event of Budget
		Populate OPEX budget accounts in the main table for better reporting and to avoid issues with child table data retrieval in case of large number of rows
	'''
	doc.accounts = []
	accounts_map = {}

	#Process OPEX budget accounts
	for row in doc.budget_accounts:
		#Set Company Currency in each row for reference during currency conversion in case company currency is not INR
		if doc.company_currency:
			row.company_currency = doc.company_currency

		# Accumulate amounts per account
		account = row.account
		# Initialize account if not exists
		if account not in accounts_map:
			accounts_map[account] = {
				'account': account,
				'budget_amount': 0
			}
			# Initialize all months
			for month in month_fields:
				accounts_map[account][month] = 0
		# Add monthly values
		for month in month_fields:
			month_value = row.get(month) or 0
			accounts_map[account][month] += month_value
			accounts_map[account]['budget_amount'] += month_value

	#Update accumulated amounts for each account in main table
	for account_data in accounts_map.values():
		doc.append('accounts', account_data)

def convert_currency(doc, method):
	'''
		Convert budget amounts for non-INR companies
	'''
	company_currency = frappe.db.get_value('Company', doc.company, 'default_currency')
	exchange_rate = 1

	if company_currency != 'INR':
		exchange_rate = frappe.db.get_value('Company', doc.company, 'exchange_rate_to_inr')
		if not exchange_rate:
			frappe.throw(
				f'Please set Exchange Rate from <b>{company_currency}</b> to <b>INR</b> for <b>{doc.company}</b>',
				title='Message',
			)

	months = [
		'january', 'february', 'march', 'april', 'may', 'june',
		'july', 'august', 'september', 'october', 'november', 'december'
	]

	def apply_conversion(row):
		'''
			Apply exchange rate conversion to a budget row
		'''
		row.budget_amount_inr = row.budget_amount * exchange_rate
		for month in months:
			setattr(row, f'{month}_inr', (getattr(row, month, 0) or 0) * exchange_rate)

	for row in (*doc.accounts, *doc.budget_accounts):
		apply_conversion(row)


def update_budget_against(doc, method=None):
	'''
		Set budget_against field based on budget_for selection
	'''
	if doc.budget_for:
		doc.budget_against = doc.budget_for
